### ISSUE TITLE
Regression: multiple temporal `CAST` (and/or `EXTRACT`) ops failing in 0.9.x vs 0.8 series

### ISSUE THREAD

**Issue by alexander-beedie** [duckdb/duckdb#9956]
### What happens?

Several previously valid casts (or extract ops) now fail, despite looking quite reasonable (eg: casting from datetime to date, or extracting the epoch from a datetime). 

(This is currently a blocker for us as we'd love to make the upgrade to take advantage of the recent list-op fixes/updates, but have a significant reliance on basic temporal casts).

### To Reproduce

Setup:
```python 
from datetime import datetime
import pandas as pd
import duckdb as dd

values = [
  datetime(2011,1,11,1,2,3), 
  datetime(2022,2,22,4,5,6),
]
df = pd.DataFrame({"dt": values})
dconn = dd.connect()
```
* Failing `datetime → date` cast:
  ```python 
  dconn.execute("SELECT dt::date FROM df").fetchall()
  # Conversion Error: 
  #   Unimplemented type for cast (TIMESTAMP_NS -> DATE)
  ```
  > Previously returned `[(date(2011,1,11),), (date(2022,2,22),)]`

* Failing `datetime → time` cast:
  ```python
  dconn.execute("SELECT dt::time FROM df").fetchall()
  # Conversion Error:
  #   Unimplemented type for cast (TIMESTAMP_NS -> TIME)
  ```
  > Previously returned `[(time(1,2,3),), (time(4,5,6),)]`

* Failing extraction of `epoch` from datetimes:
  ```python
  dconn.execute("SELECT EXTRACT(epoch FROM dt) FROM df").fetchall()
  # Binder Error: 
  #   epoch can only take temporal arguments
  ```
  > Previously returned `[(1294707723,), (1645502706,)]`

  This last one can be temporarily worked around by explicitly casting with `dt::timestamp`, but still represents a regression as this was not previously required and "dt" clearly _is_ a temporal field ;)

### OS:

macOS 14.1.2 (Apple Silicon)

### DuckDB Version:

0.9.2

### DuckDB Client:

Python

### Full Name:

Alexander Beedie

### Affiliation:

ADIA

### Have you tried this on the latest `main` branch?

I have tested with a release build (and could not test with a main build)

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by Mause** [comment]
It looks like your first and last examples are already fixed in 0.9.3-dev (though I'm aware these aren't available on pypi due to build failures)

The middle one can be worked around by first casting to `datetime` then `time`

**Comment by alexander-beedie** [comment]
> It looks like your first and last examples are already fixed in 0.9.3-dev (though I'm aware these aren't available on pypi due to build failures)

That's great to hear, thanks; will keep a close eye out for `0.9.3` landing :)
